tests/miral/reloading_config_file.cpp

```diff
@@ -0,0 +1,357 @@
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "miral/test_server.h"
+#include "miral/reloading_config_file.h"
+
+#include <wayland_wrapper.h>
+#include <gmock/gmock-function-mocker.h>
+
+#include <format>
+#include <fstream>
+
+using miral::ReloadingConfigFile;
+
+namespace
+{
+char const* const no_such_file = "no/such/file";
+char const* const a_file = "/tmp/test_reloading_config_file/a_file";
+std::filesystem::path const config_file = "test_reloading_config_file.config";
+
+class PendingLoad
+{
+public:
+    void mark_pending()
+    {
+        std::lock_guard lock{mutex};
+        pending_loads = true;
+    }
+
+    void wait_for_load()
+    {
+        std::unique_lock lock{mutex};
+
+        if (!cv.wait_for(lock, std::chrono::milliseconds{1000}, [this] { return !pending_loads; }))
+        {
+            std::cerr << "wait_for_load() failed" << std::endl;
+        }
+    }
+
+    void notify_load()
+    {
+        {
+            std::lock_guard lock{mutex};
+            pending_loads = false;
+        }
+
+        cv.notify_one();
+    }
+
+private:
+    std::mutex mutex;
+    std::condition_variable cv;
+    bool pending_loads = false;
+};
+
+struct TestReloadingConfigFile : miral::TestServer, PendingLoad
+{
+    TestReloadingConfigFile();
+    MOCK_METHOD(void, load, (std::istream& in, std::filesystem::path path), ());
+
+    std::optional<ReloadingConfigFile> reloading_config_file;
+
+    void write_a_file()
+    {
+        mark_pending();
+        std::ofstream file(a_file);
+        file << "some content";
+    }
+
+    void write_config_in(std::filesystem::path path)
+    {
+        mark_pending();
+        std::ofstream file(path/config_file);
+        file << "some content";
+    }
+
+    void SetUp() override
+    {
+        miral::TestServer::SetUp();
+        ON_CALL(*this, load(testing::_, testing::_)).WillByDefault([this]{ notify_load(); });
+    }
+};
+
+char const* const home = "/tmp/test_reloading_config_file/home";
+char const* const home_config = "/tmp/test_reloading_config_file/home/.config";
+char const* const xdg_conf_home = "/tmp/test_reloading_config_file/xdg_conf_dir_home";
+char const* const xdg_conf_dir0 = "/tmp/test_reloading_config_file/xdg_conf_dir_zero";
+char const* const xdg_conf_dir1 = "/tmp/test_reloading_config_file/xdg_conf_dir_one";
+char const* const xdg_conf_dir2 = "/tmp/test_reloading_config_file/xdg_conf_dir_two";
+
+TestReloadingConfigFile::TestReloadingConfigFile()
+{
+    std::filesystem::remove_all("/tmp/test_reloading_config_file/");
```

Comment:
I'd rather this be a generated temp directory, but I've been down this rabbit hole a bit recently…

We are leaking it, though?

Reply:
We are leaking, yes. That is intentional: when tests fail it is far easier to examine what happened. (And it is small, on a temporary filesystem)